### PR TITLE
additional explanation on the order of operations for hello world

### PR DESCRIPTION
**Pull request by 0xTutturu**
I've attempted to give a thorough explanation on why we use this specific order of operations, I hope it's clear enough and not too wordy. I've also corrected the stack order in one of the example comments since it stated "0x40" although "0x2d" was pushed to the stack.

### CODE REVIEW

File: src/tutorial/hello-world/README.md [hunk 1]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+Encoding dynamic values takes alot of work!! In order to return `"Hello, world!"` we must return 96 bytes!
+
+You might think that we can populate the memory in sequential order starting with the location (0x00), then length (0x20), and then the value (0x40), however, this would result in the value "Hello, world!" being left padded, so we would get: 
+```
+[Byte number]   [DATA]    
+0x00            0000000000000000000000000000000000000000000000000000000000000020 // The location of the "Hello, world!" data (dynamic type).
+0x20            000000000000000000000000000000000000000000000000000000000000000d // The length of "Hello, world!" in bytes
+0x40            0000000000000000000000000000000000000048656c6c6f2c20776f726c6421 // Wrong encoding!
+```
+And that's not what we need. Instead, we're going to leverage what we know about memory to produce the right result, and that is:
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:32]:
```suggestion
This is not what we need. Instead, we're going to leverage what we know about memory to produce the right result:
```

File: src/tutorial/hello-world/README.md [hunk 2]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
 0x40            48656c6c6f2c20776f726c642100000000000000000000000000000000000000 // Value "Hello, world!"
 ```
 
-Encoding dynamic values takes alot of work!! In order to return `"Hello, world!"` we must return 96 bytes!.
+Encoding dynamic values takes alot of work!! In order to return `"Hello, world!"` we must return 96 bytes!
+
+You might think that we can populate the memory in sequential order starting with the location (0x00), then length (0x20), and then the value (0x40), however, this would result in the value "Hello, world!" being left padded, so we would get:
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:25]:
```suggestion
You might think that we can populate the memory in sequential order starting with the location (0x00), then length (0x20), and then the value (0x40). However, this would result in the value "Hello, world!" being left padded. We would get: 
```

File: src/tutorial/hello-world/README.md [hunk 3]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+```
+[Byte number]   [DATA]    
+0x00            0000000000000000000000000000000000000000000000000000000000000020 // The location of the "Hello, world!" data (dynamic type).
+0x20            000000000000000000000000000000000000000000000000000000000000000d // The length of "Hello, world!" in bytes
+0x40            0000000000000000000000000000000000000048656c6c6f2c20776f726c6421 // Wrong encoding!
+```
+And that's not what we need. Instead, we're going to leverage what we know about memory to produce the right result, and that is:
+- Memory is always expanded in 32 byte increments
+- We can store a value starting from any index
+- We want to store the location starting at 0x00 and the length starting at 0x20
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:35]: May be worth separating out this one onto a new line, the other 3 bullets are facts while this one is a circumstance.

File: src/tutorial/hello-world/README.md [hunk 4]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+- Values are left padded
+
+Armed with that knowledge, we can store the value "Hello, world!" at 0x2d (starting index of length 0x20 + the length of the data in bytes 0x0d). That will give us the first 0x2d (decimal 45) bytes as zeroes due to memory expansion, and since the value is left padded and of length 0x0d (decimal 13) bytes we get another 19 bytes of zeroes, equalling to 64 bytes.
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:38]:
```suggestion
Armed with this knowledge, we can store the value "Hello, world!" at 0x2d (starting index of length 0x20 + the length of the data in bytes 0x0d). The first 0x2d (decimal 45) bytes are set to zero as a result of memory expansion. The left padding of "Hello, world!" is stored from bytes 0x2d (decimal 45)-> 0x3f (decimal 63).
```

File: src/tutorial/hello-world/README.md [hunk 5]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+```
+And that's not what we need. Instead, we're going to leverage what we know about memory to produce the right result, and that is:
+- Memory is always expanded in 32 byte increments
+- We can store a value starting from any index
+- We want to store the location starting at 0x00 and the length starting at 0x20
+- Values are left padded
+
+Armed with that knowledge, we can store the value "Hello, world!" at 0x2d (starting index of length 0x20 + the length of the data in bytes 0x0d). That will give us the first 0x2d (decimal 45) bytes as zeroes due to memory expansion, and since the value is left padded and of length 0x0d (decimal 13) bytes we get another 19 bytes of zeroes, equalling to 64 bytes. 
+
+This means our actual important bytes start at 0x40, but since memory is expanded in 32 byte increments we get right padded zeroes until 0x60, equalling 96 bytes, exactly what we need!
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:40]:
```suggestion
This means our important bytes (`48656c6c6f2c20776f726c6421") start at 0x40. However, since memory is expanded in 32 byte increments empty memory will be expanded with zeros to byte 0x60 (decimal 96), exactly what we need! 
```

File: src/tutorial/hello-world/README.md [hunk 6]
```diff
@@ -20,7 +20,32 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+
+This means our actual important bytes start at 0x40, but since memory is expanded in 32 byte increments we get right padded zeroes until 0x60, equalling 96 bytes, exactly what we need! 
+```
+[Byte number]   [DATA]    
+0x00            0000000000000000000000000000000000000000000000000000000000000000 // Empty
+0x20            0000000000000000000000000000000000000000000000000000000000000000 // Empty
+0x40            48656c6c6f2c20776f726c642100000000000000000000000000000000000000 // Correct encoding 
+```
+
+After this all we need to do is store the length of the data in 0x20, and the location of the data in 0x00 and we have what we want! Now let's see what that looks like in code.
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:48]:
```suggestion
Now we have "Hello, world!" stored, we have two things left to do. Store the length of the string in 0x20, and a pointer to the start of our dynamic data 0x00! Now let's see what that looks like in code.
```

File: src/tutorial/hello-world/README.md [hunk 7]
```diff
@@ -29,7 +54,7 @@ The following `MAIN` macro steps through this encoding in a clear way (gas optim
     // Store string "Hello, world1" in memory at 0x40
     // 0x2d is listed as mstore pads the value
     0x48656c6c6f2c20776f726c6421        // ["Hello, world!"]
-    0x2d                                // ["Hello, world!", 0x40]
+    0x2d                                // ["Hello, world!", 0x2d]
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:57]: Good catch!

**Review by cheethas — COMMENTED** [review]
Awesome work man!!! Great first contribution! You've explained the concept really well.

I've added a few wording suggestions - let me know what you think

File: src/tutorial/hello-world/README.md (hunk 3 above)
Comment (0xTutturu) [reply src/tutorial/hello-world/README.md:35]: You're right, makes more sense to have it separated.

**Comment by 0xTutturu** [comment]
> Awesome work man!!! Great first contribution! You've explained the concept really well.
> 
> I've added a few wording suggestions - let me know what you think! @0xTutturu

Thanks man! I think all of your wording suggestions are an improvement, thank you for taking the time to review the PR.

File: src/tutorial/hello-world/README.md [hunk 8]
```diff
@@ -20,7 +20,33 @@ Each part will look as follows for the string `"Hello, world!"`:
...
+This is not what we need. Instead, we're going to leverage what we know about memory to produce the right result:
+- Memory is always expanded in 32 byte increments
+- We can store a value starting from any index
+- Values are left padded
+
+Additionaly, in this example, we know that we want to store the location starting at 0x00 and the length starting at 0x20.
+
+Armed with this knowledge, we can store the value "Hello, world!" at 0x2d (starting index of length 0x20 + the length of the data in bytes 0x0d). The first 0x2d (decimal 45) bytes are set to zero as a result of memory expansion. The left padding of "Hello, world!" is stored from bytes 0x2d (decimal 45)-> 0x3f (decimal 63).
+
+This means our important bytes (`48656c6c6f2c20776f726c6421") start at 0x40. However, since memory is expanded in 32 byte increments empty memory will be expanded with zeros to byte 0x60 (decimal 96), exactly what we need!
```
Comment (cheethas) [inline src/tutorial/hello-world/README.md:41]:
```suggestion
This means our important bytes (`48656c6c6f2c20776f726c6421`) start at 0x40. However, since memory is expanded in 32 byte increments empty memory will be expanded with zeros to byte 0x60 (decimal 96), exactly what we need! 
```

**Review by cheethas — CHANGES_REQUESTED** [review]
Quick fix converting " -> `. Other than that LGTM

**Review by cheethas — APPROVED** [review]

**Review by devtooligan — APPROVED** [review]
This is great, makes the whole thing a lot clearer. Thanks fren.  I encourage you to contribute more to the docs.
